### PR TITLE
Nested Modules and SVG Image Generation in CLI

### LINKED ISSUE #26: Panic on deleting resource from tf file
Rover fails with panic if I delete a resource from .tf file:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7e222f]

goroutine 1 [running]:
main.GenerateResourceOverview(0xc000076de0, 0x1)
        /home/lera/go/src/github.com/im2nguyen/rover/rso.go:125 +0xf2f
main.generateAssets(0x89c21b, 0x5, 0x7fff859df85e, 0x32, 0x7fff859df83f, 0x12, 0xc0000eca70, 0x1, 0x1, 0xc0000eca80, ...)
        /home/lera/go/src/github.com/im2nguyen/rover/main.go:114 +0x2b9
main.main()
        /home/lera/go/src/github.com/im2nguyen/rover/main.go:61 +0x625
```

The problem is the absence of an entry for deleted resource in `map[string]*ResourceOverview`. It's created for each resource in a plan, but if it was deleted then it appears only in changes.
Anyway, creating this entry before saving a corresponding change just saves from an error. Rover won't create a node for it in the resulting overview.

### LINKED ISSUE #46: [Feature Request] A way to create the PNG only.
Hi,

Our usecase is to update the documentation with a static image only.
So ideally the rover would create the png and we can archive it in our gitlab CI pipeline.

Ive tried with the static assets zip file but it does no appear to be included.

Interactive console is not required.

### LINKED ISSUE #48: Deleted resources are not visualised at all
As part of testing Rover, I crafted a pull request for our Terraform repository that removes some resources and changes some parameters. But to my surprise, the output of Rover does not show the deleted resources at all. Am I doing wrong, or is this a known behaviour of Rover? I am using v0.2.2, for the record.

My pull request looks like this:

```
$ git diff --staged
diff --git a/terraform/clusters/dev-us-central-0/main.tf b/terraform/clusters/dev-us-central-0/main.tf
index 937d1cc6a8..ddc03aabe9 100644
--- a/terraform/clusters/dev-us-central-0/main.tf
+++ b/terraform/clusters/dev-us-central-0/main.tf
@@ -47,16 +47,6 @@ module "node_pool_main-n2s8-1" {
   }
 }

-module "node_pool_preemptible-n2s8-1" {
-  source         = "../../modules/gke/node_pool"
-  name           = "preemptible-n2s8-1"
-  cluster        = module.cluster.name
-  region         = local.region
-  machine_type   = "n2-standard-8"
-  preemptible    = true
-  max_node_count = 25
-}
-
 // node pool for GEM performance testing, not meant to
 // be utilized permanently
 module "node_pool_perf-n2s16-1" {
@@ -83,7 +73,7 @@ module "node_pool_perf-n2s8-1" {
   cluster        = module.cluster.name
   region         = local.region
   machine_type   = "n2-standard-8"
-  max_node_count = 10
+  max_node_count = 20
   taints = [
     {
       key    = "type"
@@ -186,16 +176,6 @@ module "insight-logs-gateway-dev-us-central-0-dns" {
   ttl    = 60
 }

-// dns record for amixr-03 dev environment.
-// It is needed for the services using ingress.
-module "amixr-03-dev-us-central-0-dns" {
...
-}
-
 module "network_data" {
   source = "../../modules/network_data"
 }
```

The output from Terraform looks like this:

```
Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # module.insight-logs-gateway-dev-us-central-0-dns.ns1_record.main has been changed
  ~ resource "ns1_record" "main" {
        id                = "6149ec28b8d8b600afdf141c"
      + meta              = {}
        # (5 unchanged attributes hidden)

      ~ answers {
          + meta   = {}
            # (1 unchanged attribute hidden)
        }
    }
  # module.cluster.google_container_cluster.main has been changed
  ~ resource "google_container_cluster" "main" {
        id                          = "projects/..."
        name                        = "dev-us-central-0"
        # (28 unchanged attributes hidden)










      ~ node_pool {
            name                = "main-n2s8-1"
          ~ node_count          = 98 -> 129
            # (5 unchanged attributes hidden)





            # (5 unchanged blocks hidden)
        }
      ~ node_pool {
            name                = "preemptible-n2s8-1"
          ~ node_count          = 0 -> 2
            # (5 unchanged attributes hidden)





            # (5 unchanged blocks hidden)
        }






        # (17 unchanged blocks hidden)
    }
  # module.node_pool_preemptible-n2s8-1.google_container_node_pool.main has been changed
  ~ resource "google_container_node_pool" "main" {
        id                  = "projects/..."
        name                = "preemptible-n2s8-1"
      ~ node_count          = 0 -> 2
        # (8 unchanged attributes hidden)




        # (4 unchanged blocks hidden)
    }
  # module.node_pool_main-n2s8-1.google_container_node_pool.main has been changed
  ~ resource "google_container_node_pool" "main" {
        id                  = "projects/..."
        name                = "main-n2s8-1"
      ~ node_count          = 98 -> 129
        # (8 unchanged attributes hidden)




        # (4 unchanged blocks hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions
to undo or respond to these changes.

───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place
  - destroy

Terraform will perform the following actions:

  # module.amixr-03-dev-us-central-0-dns.ns1_record.main will be destroyed
  - resource "ns1_record" "main" {
      - id                = "611b910118fbde00a7c861d5" -> null
      ...
    }

  # module.node_pool_perf-n2s8-1.google_container_node_pool.main will be updated in-place
  ~ resource "google_container_node_pool" "main" {
        id                  = "projects/..."
        name                = "perf-n2s8-1"
        # (9 unchanged attributes hidden)

      ~ autoscaling {
          ~ max_node_count = 10 -> 20
            # (1 unchanged attribute hidden)
        }



        # (3 unchanged blocks hidden)
    }

  # module.node_pool_preemptible-n2s8-1.google_container_node_pool.main will be destroyed
  - resource "google_container_node_pool" "main" {
      - cluster             = "dev-us-central-0" -> null
      ...
    }

Plan: 0 to add, 1 to change, 2 to destroy.

───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

Any ideas why Rover does not show the deleted resources?

### LINKED ISSUE #57: nested modules not shown in rover.
Great project first of all ! I have a tf state with main module in the main.tf/seed.tf . This module utilizes/depends on other modules so there are nested modules. Once I do a plan I can see it generated in tf plan (section of plan file here) : 

```
  # module.search-query.aws_route53_record.route53 will be created
  + resource "aws_route53_record" "route53" {
      + allow_overwrite = (known after apply)
      + fqdn            = (known after apply)
      + id              = (known after apply)
      + name            = ""
      + records         = (known after apply)
      + ttl             = 60
      + type            = "CNAME"
      + zone_id         = ""
    }

  # module.search-query.module.acm.module.acm_certificate.aws_route53_record.validation[0] will be created
  + resource "aws_route53_record" "validation" {
      + allow_overwrite = true
      + fqdn            = (known after apply)
      + id              = (known after apply)
      + name            = (known after apply)
      + records         = (known after apply)
      + ttl             = 300
      + type            = (known after apply)
      + zone_id         = ""
    }

  # module.search-query.module.acm.module.acm_certificate.aws_route53_record.validation[1] will be created
  + resource "aws_route53_record" "validation" {
      + allow_overwrite = true
      + fqdn            = (known after apply)
      + id              = (known after apply)
      + name            = (known after apply)
      + records         = (known after apply)
      + ttl             = 300
      + type            = (known after apply)
      + zone_id         = ""
    }

  # module.search-query.module.acm.module.acm_certificate.aws_route53_record.validation[2] will be created
  + resource "aws_route53_record" "validation" {
      + allow_overwrite = true
      + fqdn            = (known after apply)
      + id              = (known after apply)
      + name            = (known after apply)
      + records         = (known after apply)
      + ttl             = 300
      + type            = (known after apply)
      + zone_id         = ""
    }

Plan: 13 to add, 0 to change, 0 to destroy.
```

^ here, you see how the first change is the main module but the rest are nested modules and all of them are going to be created. Im running rover on this : 

```
(venv) ➜  virginia git:(master) ✗ ./rover_v0.2.2 -planPath plan.out -tfConfigExists=false -tfPath terraform1.0 -standalone true
2021/10/15 14:52:04 Starting Rover...
2021/10/15 14:52:04 Using provided plan...
2021/10/15 14:52:06 Generating resource overview...
2021/10/15 14:52:06 Generating resource map...
2021/10/15 14:52:06 Generating resource graph...
2021/10/15 14:52:06 Done generating assets.
2021/10/15 14:52:06 Generated zip file: rover.zip
```

This creates the following : 
Screen Shot 2021-10-15 at 2 59 00 PM

where you can see the main module and the resources created in the module directly but not the nested module. What is interesting is that it knows about it though as if I scroll down to resources, I can see it being listed just not displayed in graph ( you can see the nested modules at the end of the main module -  search-query)  :

Screen Shot 2021-10-15 at 2 59 13 PM

### LINKED ISSUE #60: Issue: Save Graph not working
First off, this is a great tool! 

Im trying to save the output to present in documentation etc etc 
When the image is downloaded, it says its always empty.
Tried different browsers etc and no change. 

Can workaround using straight screenshots of the display. 

Bonus : Could we get a full screen mode of the actual visualization ? i.e remove the legend and what not so we can move around large TF code 

Platform : OSX
Version : rover_0.2.2_darwin_amd64

### PR DESCRIPTION
**Pull request by JackFlukinger**
This is a big one.

Adds:

- Full nested module support
- RSO rewritten to support nested modules and spotty configuration
- Graph node and edge refactor to generate from resource map instead of going back through plans, states, configs again
    - Can make changes to graph output by modifying map
- SVG instead of PNG output -- browsers only allow downloaded images up to some unspecified size. I had some trouble with large graphs not exporting due to this. SVG format makes the images much smaller, and infinitely scalable
- Some map updates, always shows line numbers (even in nested modules) if available
    - Nested resources and modules are indented for general readability
- Terraform 1.1.2 support
- SVG export support in CLI with `-genImage true`. Probably does not work with docker, though.

I'm happy to go into a ton more detail, but it all works. If you don't want to PR because this is so ginormous I get it.

Resolves #60
Resolves #57 
Resolves #48 
Resolves #46 
Resolves #26

### CODE REVIEW

**Comment by EugenMayer** [comment]
Is this PR already considered sind of working? For now, since our project heavily uses nested modules, rover is yet unusable and i would love getting started with it. Any hints?

**Comment by JackFlukinger** [comment]
> Is this PR already considered sind of working? For now, since our project heavily uses nested modules, rover is yet unusable and i would love getting started with it. Any hints?

Give it a try! I’ve tested pretty extensively on my own projects. Let me know if it works for you.

**Comment by EugenMayer** [comment]
Considering the size of the PR, are there even any plans to merge it @im2nguyen ? Just curious to understand the direction of the project

**Comment by im2nguyen** [comment]
Yep! Sorry, wasn't aware about this. I'll review tomorrow!!

Thanks for getting these features fleshed out!

**Comment by EugenMayer** [comment]
Thank you @im2nguyen and of cours @JackFlukinger for all the hard work

**Review by im2nguyen — APPROVED** [review]
Overall this looks great! Thank you for tackling this @JackFlukinger!! 🎉 

I created a PR on your branch that allows it to work with the Docker image. Once you take a look at that, I'll be more than happy to merge this 😄 

I added a small blurb about image generation to the README.md -- we need to update the README to raise awareness of the other flags too... but that can be a future update

File: graph.go [hunk 1]
```diff
@@ -86,218 +88,146 @@ func (r *rover) GenerateGraph() error {
...
+				pid = strings.TrimSuffix(pid, nodeMap[parent].Data.Label)
+				pid = strings.TrimSuffix(pid, ".")
+			}
+
+			label := strings.TrimPrefix(id, pid)
+			label = strings.TrimPrefix(label, ".")
+			label = strings.TrimPrefix(label, "module.")
+			label = strings.TrimPrefix(label, "local.")
+			label = strings.TrimPrefix(label, "output.")
+			label = strings.TrimPrefix(label, "var.")
```
[the remark this answers is not in the document]
Comment (JackFlukinger) [reply graph.go]: Just committed that. But yeah, that block deals with every "resource" in the map (confusing term) that isn't either a file or proper terraform _resource_. So locals, modules, outputs, vars, etc.

**Comment by JackFlukinger** [comment]
Thanks so much @im2nguyen ! Should be good to go.

**Comment by im2nguyen** [comment]
Merged! Can you send me your email address @JackFlukinger? I can send you some Terraform swag if you're interesting 😄

**Comment by JackFlukinger** [comment]
Shot you a friend request on LinkedIn @im2nguyen w/ my email. Thanks -- I'd love some!

**Pull request merged**